### PR TITLE
Introduce v3 backend maps

### PR DESCRIPTION
**Pull request by YutaroHayakawa**
Introduce new v3 backend maps for IPv4 and IPv6. Since it modifies the value size of the backend maps, this PR contains the map migration logic as well. The change from the previous map version now includes the `cluster_id` field. This field will be filled by the control plane that implements Cluster Mesh with overlapping PodCIDR support. When the global service contains backends from the remote clusters with overlapping PodCIDR, we can distinguish them by cluster-aware addressing scheme (see https://github.com/cilium/cilium/pull/21161). Note that since the current agent doesn't have actual logic to realize global service with cluster-aware addressing, `cluster_id` should always be zero for now.

```release-note
Introduce v3 backend maps
```

### CODE REVIEW

**Comment by brb** [comment]
Have we decided to have or not to have the V3 -> V2 downgrade path?

**Comment by YutaroHayakawa** [comment]
Yep, we should have it. Should be covered with separate PRs.

**Review by ti-mo — APPROVED** [review]
Looks okay from my side given up/downgrade path is added later.

Just curious, why do we bother renaming `LB6_BACKEND_MAP_V2` to `LB6_BACKEND_MAP_V3` if it's a macro? Updating the `cDefinesMap` entry should be sufficient.

**Comment by ldelossa** [comment]
Small nit, but the first two commits could probably be combined into one. They are both small and edit the same file.

**Review by ldelossa — APPROVED** [review]
LGTM.

**Comment by YutaroHayakawa** [comment]
Squashed the first two commits, as Louis mentioned. Also, I noticed third commit was not used anywhere at all. I mistakenly ported it from my dev branch, so I dropped the third commit.

File: pkg/maps/lbmap/ipv4.go
Comment (jrajahalme) [inline pkg/maps/lbmap/ipv4.go:545]: It would be nice if the version indicators in key and value would be in sync with the map version, so maybe rename as `Backend4KeyV2` and `Backend4ValueV3` (and the same for IPv6 versions)?

Comment (YutaroHayakawa) [reply pkg/maps/lbmap/ipv4.go:545]: Ok, I'll fix.

Comment (YutaroHayakawa) [reply pkg/maps/lbmap/ipv4.go:545]: Fixed

**Comment by YutaroHayakawa** [comment]
Fixed naming stuff.

**Comment by YutaroHayakawa** [comment]
LB conformance test is a regression. Let me fix that.

test-1.16-4.9 is failing on the downgrading path, which is expected. Maybe I should merge the downgrading changes to stable branches first for let this PR merged?

**Comment by YutaroHayakawa** [comment]
Fixed NAT64/46 regression. The reason was the new `NewBackend6ValueV3` constructor didn't consider the case it is used for string IPv4 backend for NAT64/46 case.

https://github.com/cilium/cilium/blob/b9f72929d45ba5d410ae7181c9db6770da4948e3/pkg/maps/lbmap/ipv6.go#L294-L297

The interesting thing here was the old constructor `NewBackend6Value` "checks" whether the input is an IPv6 address or not, but it was not working because `To16()` returns without an error even if the input is an IPv4 address. So, maybe it has been working for NAT64/46 accidentally?

https://github.com/cilium/cilium/pull/21797/files#diff-6884fdfcb2018c025115daa75f477e513b8f9bfc6320f35520e697d6ae5421c5R356-R359

**Comment by YutaroHayakawa** [comment]
/test-1.24-5.4

**Comment by YutaroHayakawa** [comment]
/test-1.25-4.19

**Comment by YutaroHayakawa** [comment]
/test-1.26-net-next

**Comment by YutaroHayakawa** [comment]
Once we merge https://github.com/cilium/cilium/pull/22416 and release v.1.12.x, we can test this branch again. So that we can try upgrade/downgrade works correctly or not.

**Comment by YutaroHayakawa** [comment]
https://github.com/cilium/cilium/pull/22416 merged. And I tweaked the test code to checkout the helm chart of that commit.

**Comment by YutaroHayakawa** [comment]
/test-1.16-4.9

**Review by nebril — APPROVED** [review]
LGTM, just make sure to remove the "Don't Merge" commit ;)

**Comment by YutaroHayakawa** [comment]
/test-1.16-4.9

**Comment by YutaroHayakawa** [comment]
/test-1.16-4.9

**Comment by YutaroHayakawa** [comment]
/test-1.16-4.9

**Comment by YutaroHayakawa** [comment]
Ok, the Upgrade/Downgrade test passed. I'll do run rest of the tests and if they pass, I'll revert the `Don't merge` commit and make this ready-to-merge.

**Comment by YutaroHayakawa** [comment]
Some tests failed with `Error: Unable to install Cilium: roles.rbac.authorization.k8s.io "cilium-config-agent" already exists`. Perhaps this is related to the recent reverting of the per-node configuration feature. I'll try with rebased tree.

**Comment by YutaroHayakawa** [comment]
/test-1.25-4.19

**Comment by YutaroHayakawa** [comment]
Ok, now it's all green. I'll revert the changes for testing and push again, but since it passed all tests, I'll make this ready to merge.

スクリーンショット 2022-12-10 0 49 12

**Comment by pchaigno** [comment]
Seeing a lot of CI job reruns without justifications here :grimacing:

**Pull request merged**
